src/librbd/Lock.cc

```diff
@@ -0,0 +1,620 @@
+                        thread_pool_singleton, "librbd::lock::thread_pool");
+
+  m_work_queue = new ContextWQ("librbd::lock::op_work_queue",
+                               m_cct->_conf->rbd_op_thread_timeout,
+                               thread_pool_singleton);
+
+  m_watcher = new L(this);
+}
+
+template <typename L>
+Lock<L>::~Lock() {
+  {
+    Mutex::Locker locker(m_lock);
+    assert(m_state == STATE_SHUTDOWN || (
+           m_state == STATE_UNLOCKED && !m_watcher->is_registered()));
+
+    delete m_watcher;
+
+    if (m_policy != nullptr) {
+      delete m_policy;
+    }
+  }
+
+  m_work_queue->drain();
+  delete m_work_queue;
+}
+
+template <typename L>
+bool Lock<L>::is_lock_owner() const {
+  Mutex::Locker locker(m_lock);
+
+  bool lock_owner;
+
+  switch (m_state) {
+  case STATE_LOCKED:
+  case STATE_REACQUIRING:
+  case STATE_PRE_SHUTTING_DOWN:
+    lock_owner = true;
+    break;
+  default:
+    lock_owner = false;
+    break;
+  }
+
+  ldout(m_cct, 20) << this << " " << __func__ << "=" << lock_owner
+                   << dendl;
+  return lock_owner;
+}
+
+template <typename L>
+void Lock<L>::shut_down(Context *on_shut_down) {
+  ldout(m_cct, 10) << this << " " << __func__ << dendl;
+
+  {
+    Mutex::Locker locker(m_lock);
+    assert(!is_shutdown_locked());
+    execute_action(ACTION_SHUT_DOWN, on_shut_down);
+  }
+
+  // if stalled in request state machine -- abort
+  handle_peer_notification();
+}
+
+template <typename L>
+void Lock<L>::try_lock(Context *on_tried_lock) {
+  int r = 0;
+  {
+    Mutex::Locker locker(m_lock);
+    if (is_shutdown_locked()) {
+      r = -ESHUTDOWN;
+    } else if (m_state != STATE_LOCKED || !m_actions_contexts.empty()) {
+      ldout(m_cct, 10) << this << " " << __func__ << dendl;
+      execute_action(ACTION_TRY_LOCK, on_tried_lock);
+      return;
+    }
+  }
+
+  on_tried_lock->complete(r);
+}
+
+template <typename L>
+void Lock<L>::request_lock(Context *on_locked) {
+  int r = 0;
+  {
+    Mutex::Locker locker(m_lock);
+    if (is_shutdown_locked()) {
+      r = -ESHUTDOWN;
+    } else if (m_state != STATE_LOCKED || !m_actions_contexts.empty()) {
+      ldout(m_cct, 10) << this << " " << __func__ << dendl;
+      execute_action(ACTION_REQUEST_LOCK, on_locked);
+      return;
+    }
+  }
+
+  if (on_locked != nullptr) {
+    on_locked->complete(r);
+  }
+}
+
+template <typename L>
+void Lock<L>::release_lock(Context *on_released) {
+  int r = 0;
+  {
+    Mutex::Locker locker(m_lock);
+    if (is_shutdown_locked()) {
+      r = -ESHUTDOWN;
+    } else if (m_state != STATE_UNLOCKED || !m_actions_contexts.empty()) {
+      ldout(m_cct, 10) << this << " " << __func__ << dendl;
+      execute_action(ACTION_RELEASE_LOCK, on_released);
+      return;
+    }
+  }
+
+  on_released->complete(r);
+}
+
+template <typename L>
+void Lock<L>::reacquire_lock(Context *on_reacquired) {
+  {
+    Mutex::Locker locker(m_lock);
+
+    if (!is_shutdown_locked() &&
+               (m_state == STATE_LOCKED ||
```

Comment:
Minor: spacing looks odd in the review tool -- not sure if this is indented past ```!is_shutdown_locked()```, but if it is, it should be moved back